### PR TITLE
[grpc] deferred snapshot support

### PR DESCRIPTION
**Pull request by alangenfeld**
Adds a `defer_snapshots` arg to `ExternalRepository` and a new grpc method `ExternalJob` for loading the data/snapshot for single job.

origin prototype https://github.com/dagster-io/dagster/pull/9563

### How I Tested These Changes

added test

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* master
  * **PR #9613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9613" target="_blank">Graphite</a> 
    * **PR #9627** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9627" target="_blank">Graphite</a> 
      * **PR #9630** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9630" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/9630?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/host_representation/external.py [hunk 1]
```diff
@@ -216,7 +272,12 @@ def solid_names_in_topological_order(self):
 
     @property
     def external_pipeline_data(self):
-        return self._external_pipeline_data
+        if self._data is None:
+            if self._ref is None or self._ref_to_data_fn is None:
+                check.failed("unexpected state - unable to load data from ref")
+            self._data = self._ref_to_data_fn(self._ref)
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/host_representation/external.py:278]: worried about thread safety here - is there anywhere where two different threads call this on the same object? Before it was immutable

File: python_modules/dagster/dagster/_grpc/client.py [hunk 2]
```diff
@@ -252,10 +256,31 @@ def external_repository(self, external_repository_origin):
...
+            defer_snapshots=defer_snapshots,
         )
 
         return res.serialized_external_repository_data
 
+    def external_job(
+        self,
+        external_repository_origin: ExternalRepositoryOrigin,
+        job_name: str,
+    ) -> str:
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/client.py:268]: what's the back-compat plan here in OSS - to make this feature opt-in and if you opt-in you need to make sure everything is upgraded? or do so some kind of fallback in the future if we add it?

**Review by gibsondan — COMMENTED** [review]
this broadly makes sense to me other than the thread-safety q - do we possibly need a lock? i think the daemon at least passes a workspace around and expects to be able to call these methods on it

File: python_modules/dagster/dagster/_core/host_representation/external.py (hunk 1 above)
Comment (gibsondan) [reply python_modules/dagster/dagster/_core/host_representation/external.py:278]: maybe this won't cause issues other than possibly duplicated work if its called from two different threads? still spooky

File: python_modules/dagster/dagster/_grpc/client.py (hunk 2 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_grpc/client.py:268]: This grpc method is only called when the deferred snapshots responses comes back, so we only call it when we we know we are hitting a recent-enough grpc server. 

If an external_job call got added somewhere else that could be a problem.

File: python_modules/dagster/dagster/_core/host_representation/external.py (hunk 1 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/host_representation/external.py:278]: ya agree it shouldn't break anything but could cause repeat fetches

File: python_modules/dagster/dagster/_core/host_representation/external_data.py [hunk 3]
```diff
@@ -48,29 +48,28 @@ class ExternalRepositoryData(
         "_ExternalRepositoryData",
         [
             ("name", str),
-            ("external_pipeline_datas", Sequence["ExternalPipelineData"]),
             ("external_schedule_datas", Sequence["ExternalScheduleData"]),
             ("external_partition_set_datas", Sequence["ExternalPartitionSetData"]),
             ("external_sensor_datas", Sequence["ExternalSensorData"]),
             ("external_asset_graph_data", Sequence["ExternalAssetNode"]),
+            ("external_pipeline_datas", Optional[Sequence["ExternalPipelineData"]]),
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/host_representation/external_data.py:55]: re-ordering doesn't confuse serdes i assume

File: python_modules/dagster/dagster/_grpc/protos/api.proto [hunk 4]
```diff
@@ -168,3 +170,12 @@ message StartRunReply {
...
 }
+
+message ExternalJobRequest {
+  string serialized_repository_origin = 1;
+  string job_name = 2;
+}
+
+message ExternalJobReply {
+  string serialized_response = 1;
+}
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_grpc/protos/api.proto:181]: per feedback on other diff - could imagine having two fields, one for good, one for error, vs. parsing the string looking for clues that it's a serialized error - not the end of the world

File: python_modules/dagster/setup.py [hunk 5]
```diff
@@ -87,7 +87,7 @@ def get_version() -> str:
             "test": [
                 "coverage==5.3",
                 "docker",
-                "grpcio-tools",
+                "grpcio-tools>=1.32.0,<1.44.0",  # related to above grpcio pins
```
Comment (gibsondan) [inline python_modules/dagster/setup.py:90]: related?

**Review by gibsondan — APPROVED** [review]

File: python_modules/dagster/setup.py (hunk 5 above)
Comment (alangenfeld) [reply python_modules/dagster/setup.py:90]: ya needed this to run the proto gen script

**Pull request merged**